src/java_tools/junitrunner/java/com/google/testing/junit/runner/model/TestSuiteModel.java

```diff
@@ -183,6 +184,18 @@ public void testFailure(Description description, Throwable throwable) {
         test.testFailure(throwable, now());
       }
     }
+    else {
+      final Optional<Description> matchingSuite = testsMap.keySet().stream()
+          .filter(d -> d.isSuite() && d.getTestClass().equals(description.getTestClass()))
+          .findAny();
+      if (matchingSuite.isPresent()) {
+        final TestSuiteNode suite = (TestSuiteNode)testsMap.get(matchingSuite.get());
+        final TestCaseNode newTest = new TestCaseNode(description, suite);
+        suite.addTestCase(newTest);
+        testsMap.put(description, newTest);
+        newTest.testFailure(throwable, now());
```

Comment:
TestNode differentiates between `testFailure` and `dynamicTestFailure`. I'm not familiar with this code, but would it not be better to do this as well?

Perhaps the code should be written so the logic to get a `TestNode` object (and update the map) comes at the top, and then the original code just operates on that?

Reply:
@c-mita for this particular kind of exception, at least, when I try replacing the call to `testFailure` with `dynamicTestFailure`, I don't get a full stack trace in the final XML, despite passing through the Throwable.

So I'm a little hesitant to reorder the logic, because I don't understand the circumstances under which a DynamicTestException will appear in the top part of the if statement.

Reply:
After a little more investigation, I _believe_ that `DynamicTestException` is unused within Bazel.

It is used within Google however, so some care is needed. I'll need to investigate a bit more as changes here could have impacts beyond Bazel's test exection.

Reply:
That using `dynamicTestFailure` all the time surpresses full stack traces doesn't surprise me, because that appears to be part of its intent. Its origin lies in handling Javascript test suites (which is an area I have no expertise in).

But the code I had in mind was this:
```
if (throwable instanceof DynamicTestException) {
  DynamicTestException dynamicFailure = (DynamicTestException) throwable;
  newTest.dynamicTestFailure(dynamicFailure.getTest(), dynamicFailure.getCause(), now());
} else {
  newTest.testFailure(throwable, now());
}
```
Essentially replicating what's done in the method above.

Reply:
Moved the logic into getTest() think that's what you meant originally.

Reply:
This means this new logic will run for more than just test failures; I don't know if this is a good thing or not.